proposed/SerializerGoals5.0.md

```diff
@@ -0,0 +1,281 @@
+
+Ideally we add a high-performance emit replacement for targeted scenarios such as property setters, getters and constructors.
+
+Some serializers do not have a non-emit fallback. Although System.Text.Json has a fallback for emit, it does not have a fallback for its usage of `Type.MakeGenericType`.
+
+## Faster startup performance
+
+This is important for scenarios including [micro-services](https://github.com/dotnet/runtime/issues/1568) where processes start and exit frequently. It is also important for benchmarks including TechEmpower.
+
+The serializer has fast startup compared to most other serializers including Newtonsoft's JSON.Net, JIL and Utf8Json. Startup perf is mostly affected by generating IL - the Jil serializer is the most aggressive here and has the highest startup cost.
+
+There is no stated goal for 5.0 to make the steady-state CPU performance faster than it is now in 5.0 for general sceanrios, although we do not want to noticably regress steady-state performance while addressing the stated goals including improving startup performance. It is possible some code-gen options described later may have steady-state performance gains, but again that is not a stated goal.
+
+<details>
+  <summary>5.0 master startup benchmarks</summary>
+
+A small POCO was used (`LoginViewModel` benchmark test class).
+
+| Serializer |  Serialize (us)  | Serialize ratio | Deserialize (us) | Deserialize ratio |
+| :-- | :-- | :-- | :-- | :--
+| **System.Text.Json** | 4,720 | 1.00 | 19,379 | 1.00 
+| **Json.NET** | 24,916 | 5.28 | 102,226 | 5.28 
+| **Utf8Json** | 7,290 | 1.54 | 106,817 | 5.51 
+| **Jil** | 66,456 | 14.08 | 170,548 | 8.80 
+</details>
+
+## Reduced memory usage
+Reducing private bytes (non-shared or per-process) memory is important for Bing scenarios amongst others.
+    - This can be done by reducing emit\`Type.MakeGenericType` and by capturing POCO metadata in generated code instead of RAM.
+
+## Reduced size-on-disk
+Primarily this is for faster downloads of the runtime and applications
+
+The runtime can be made smaller by removing emit support. Some runtimes including iOS do not support dynamic code generation (emit or jitting) due to policy and thus require AOT (Ahead-Of-Time compilation).
+
+The application can be made smaller by support aggressive linking behavior (e.g. "tree-shaking"). This can be done in a number of ways outside the scope of this document. However, approaches in this document that do code-gen may automatically handle this due to generated calls to members.
+
+# Serializer design background
+Today the serializer avoids using Reflection.Emit on non-.NET scenarios through the use of an #ifdef. However, the serializer still uses emit indirectly through `Type.MakeGenericType`. Note that Xamarin stubs `MakeGenericType` and tries to support these on iOS by pre-generating types via AOT but there are still issues with this (including not being able to determine all possible `<T>` variants).
```

Comment:
> the serializer still uses emit indirectly through `Type.MakeGenericType`

This is not really emit. I would describe `Type.MakeGenericType` as requiring runtime code generation or interpreter. 

It may be useful to use Reflection.Emit instead of emit throughout the code.

Reply:
I made a pass through and removed emit-based terminology when it could also apply to any runtime code generation.